### PR TITLE
Add animation time control for sprites and 3D models

### PR DESCRIPTION
**Pull request by D8H**
- Fix 3D model animation changes from resuming animation.

### Demo
- https://gd.games/instant-builds/c550b6d3-8a14-446a-b39d-b445992ef3f5?dev=true

### CODE REVIEW

**Comment by 4ian** [comment]
See this video: 
- With the sprite, I can switch between "Jump" and "Run" and even if the "Jump" animation has ended, I can click on "Run" and it will play the animation.
- With the 3d model, I can switch between "Jump" and "Run" and:
  - If the "Jump" animation has not finished, when I click on "Run", then "Run" animation will be played.
  - If the "Jump" animation has finished, when I click on "Run", then "Run" animation will not played. I need to click on "Resume".

https://github.com/4ian/GDevelop/assets/1280130/7be546df-b547-444f-afef-c396c1be39db

Is this expected?

File: Extensions/3D/Model3DRuntimeObject3DRenderer.ts [hunk 1]
```diff
@@ -347,6 +348,27 @@ namespace gdjs {
       this._action.play();
       // Make sure the first frame is displayed.
       this._animationMixer.update(0);
+      if (wasPaused) {
+        this.pauseAnimation();
```
Comment (AlexandreSi) [inline Extensions/3D/Model3DRuntimeObject3DRenderer.ts:352]: Shouldn't this be:
```suggestion
        this.resumeAnimation();
```
?

File: Core/GDCore/Extensions/Builtin/Capabilities/AnimatableExtension.cpp [hunk 2]
```diff
@@ -130,6 +130,30 @@ void GD_CORE_API BuiltinExtensionsImplementer::ImplementsAnimatableExtension(
...
+      .MarkAsAdvanced();
+
+  aut.AddExpression(
+         "Duration",
+         _("Animation duration"),
+         _("Return the animation duration (in seconds)."),
+         _("Animations and images"),
+         "res/actions/animation24.png")
+      .AddParameter("object", _("Object"))
+      .AddParameter("behavior", _("Behavior"), "AnimatableBehavior");
```
Comment (AlexandreSi) [inline Core/GDCore/Extensions/Builtin/Capabilities/AnimatableExtension.cpp:156]: Isn't the animation name missing?

Comment (AlexandreSi) [reply Core/GDCore/Extensions/Builtin/Capabilities/AnimatableExtension.cpp:156]: Some methods in the runtime use it, some don't. I think it should be clarified

File: Extensions/3D/Model3DRuntimeObject3DRenderer.ts (hunk 1 above)
Comment (D8H) [reply Extensions/3D/Model3DRuntimeObject3DRenderer.ts:352]: It keeps the "pause" state untouched because `play` is automatically starting the animation.

Comment (4ian) [reply Extensions/3D/Model3DRuntimeObject3DRenderer.ts:352]: Worth a comment

**Comment by D8H** [comment]
> See this video:
> 
>     * With the sprite, I can switch between "Jump" and "Run" and even if the "Jump" animation has ended, I can click on "Run" and it will play the animation.
> 
>     * With the 3d model, I can switch between "Jump" and "Run" and:
>       
>       * If the "Jump" animation has not finished, when I click on "Run", then "Run" animation will be played.
>       * If the "Jump" animation has finished, when I click on "Run", then "Run" animation will not played. I need to click on "Resume".
> 
> Is this expected?

It's not expected, it's because I relied on Three.js state, I'll use our own state.

File: Core/GDCore/Extensions/Builtin/Capabilities/AnimatableExtension.cpp (hunk 2 above)
Comment (D8H) [reply Core/GDCore/Extensions/Builtin/Capabilities/AnimatableExtension.cpp:156]: It's the current animation duration.
Event-based extensions won't necessarily know the animation name so it seems easier this way.
I'll clarify it in the description.

**Comment by AlexandreSi** [comment]
Regarding the naming: in case we need a getter for an animation other than the current one:
- should we change the method `GetAnimationDuration` to `GetCurrentAnimationDuration` so that the name is available.
- or should we leave it this way and we would name the new method `GetAnimationDurationByName`?
I feel like the first option is clearer but I'll let you decide

**Comment by D8H** [comment]
> Regarding the naming: in case we need a getter for an animation other than the current one:
> 
> * should we change the method `GetAnimationDuration` to `GetCurrentAnimationDuration` so that the name is available.
> 
> * or should we leave it this way and we would name the new method `GetAnimationDurationByName`?
> 
> I feel like the first option is clearer but I'll let you decide

I guess if the feature is asked, we could add an optional parameter to the `Duration` expression.
JS side it would be clearer to do what you asked but it needs to match the expression signature.

**Review by AlexandreSi — APPROVED** [review]

File: GDJS/Runtime/spriteruntimeobject.ts [hunk 3]
```diff
@@ -475,40 +475,29 @@ namespace gdjs {
...
-        this._currentFrame = Math.min(
-          Math.floor(this._animationElapsedTime / direction.timeBetweenFrames),
-          direction.frames.length - 1
+        const animationElapsedTime =
+          this._animationElapsedTime +
+          (this.getElapsedTime() / 1000) * this._animationSpeedScale;
+        this.setAnimationElapsedTime(
+          direction.loop
+            ? gdjs.evtTools.common.mod(animationElapsedTime, animationDuration)
+            : gdjs.evtTools.common.clamp(
```
Comment (4ian) [inline GDJS/Runtime/spriteruntimeobject.ts:490]: In theory this clamp is not needed as done by setAnimationElapsedTime?

Comment (D8H) [reply GDJS/Runtime/spriteruntimeobject.ts:490]: Yes, it's for clarity.

**Pull request merged**